### ISSUE TITLE
Investigate why page loads are so slow

### ISSUE THREAD

**Issue by PaulHeggarty** [lingdb/IELex2-CognaC#84]
And see if this can be fixed a bit!

**Comment by runjak** [comment]
Yes, I'd appreciate faster page loads especially since it also hinders development.
I discussed some ideas on improving load times with Hans-Jörg on Monday and will see how I can implement them.

**Comment by runjak** [comment]
As a note to myself: The current usage of the `process_postrequest_form` functions in `views.py` should be considered an anti-pattern since the `WTForms` library brings this functionality itself and there is no need to replicate it and potentially introduce errors or delay with it.

**Comment by cormacanderson** [comment]
By no means a solution, but might implementing https://github.com/lingdb/IELex2-CognaC/issues/76 improve things even a little for now? Presuming this is the case, I've updated that to Urgency 0.

**Comment by runjak** [comment]
Writing with @xrotwang gave me some hints on how to look at performance problems and what typical catches might be. There appear to be a bunch of places in the code where a query is executed several times in a row but a single execution could be enough.
One example is [views.py:558-562](https://github.com/lingdb/CoBL/blob/master/ielex/views.py#L558-L562), where the MeaningListOrder query appears to be quite the same in many cases but is executed something like `197` times when viewing a single language.
Looking after such cases I could be able to find some performance problems.

**Comment by xrotwang** [comment]
Yes, it looks as if your example could be replaced by an appropriate `prefetch_related` call chained to the [query](https://github.com/lingdb/CoBL/blob/master/ielex/views.py#L540-L543).

**Comment by runjak** [comment]
There are still cases left that perform bad likely because they use neither `select_related` nor `prefetch_related`, but for the ones I've already touched I can report that the timing displayed in the Django Debug Toolbar has improved by seconds in each case. If I can solve the other cases in a similar fashion this could give the site a substantial performance improvement.

**Comment by runjak** [comment]
I've had a look at the clades table to make a few tests.
This table is generally one of the faster views, but I wanted to see how loading times for the site may change when using [gunicorn](http://docs.gunicorn.org/en/stable/deploy.html) instead of a python standalone process to serve the site. I used [boom](https://github.com/rakyll/boom) to send `500` requests to the `/clades/` route sending `1` request at a time and sending `50` requests concurrently:

#### `python manage.py runserver`, `1` request at a time:
![screenshot_2016-08-03_11-52-22](https://cloud.githubusercontent.com/assets/5417642/17361552/643ce418-5971-11e6-885c-9a6822989551.png)
#### `gunicorn --threads 4 wsgi:application`, `1` request at a time:
![screenshot_2016-08-03_11-52-35](https://cloud.githubusercontent.com/assets/5417642/17361549/64395550-5971-11e6-98db-e6a95638afb4.png)
#### `python manage.py runserver`, `50` requests at a time:
![screenshot_2016-08-03_11-52-52](https://cloud.githubusercontent.com/assets/5417642/17361550/6439efc4-5971-11e6-9cbb-c412491f137c.png)
#### `gunicorn --threads 4 wsgi:application`, 50 requests at a time:
![screenshot_2016-08-03_11-53-08](https://cloud.githubusercontent.com/assets/5417642/17361551/643a71c4-5971-11e6-978a-610058b943e7.png)

From this I gather the following conclusions:
* While performance for a single request at a time is similar in both cases it degrades for the non gunicorn solution once the number of concurrent requests grow. This is expected because the simple variant is only running a single process and must keep up to `49` requests waiting while generating the response for one of them.
* While gunicorn appears to be a good improvement performance wise it can't be expected to solve all our performance problems. When the code is slow or badly written I still need to improve it.
* I will deploy gunicorn on the server soon.

**Comment by xrotwang** [comment]
gunicorn is also very configurable; so you can tune performance even if the app is not well behaved, e.g. you could limit resources spent per worker or per request, which may be good if you have views that take a long time sporadically, only.

**Comment by xrotwang** [comment]
Any particular reason why you used gunicorn with threads instead of multiple workers? With multiple workers gunicorn can make use of multiple CPUs while with multiple threads it cannot (due to the GIL).

**Comment by runjak** [comment]
No particular reason, I was just unaware of this - I'll switch to workers then!

**Comment by runjak** [comment]
I've repeated the test with `gunicorn --workers 8  wsgi:application`:
![screenshot_2016-08-03_13-49-57](https://cloud.githubusercontent.com/assets/5417642/17364580/4171a922-5981-11e6-955c-b0cfb0d43645.png)

~6 seconds faster :)

**Comment by runjak** [comment]
I've also tried `16` workers on my machine:
![screenshot_2016-08-03_13-56-00](https://cloud.githubusercontent.com/assets/5417642/17364707/114a1f12-5982-11e6-9ef5-b56af73c35ae.png)

**Comment by xrotwang** [comment]
I think the rule of thumb is: [(CPUs x 2) + 1](http://docs.gunicorn.org/en/stable/design.html#how-many-workers).

**Comment by PaulHeggarty** [comment]
This is an excellent improvement, for what doesn’t seem to have taken up too much of your time.  However, perhaps you’ve misinterpreted what I said yesterday about Cassandra's comment on how much it would help her if page loads were speeded up.  That was meant only in the wider context of all of the issues at priority level 0, and some at level 1 as well.  

The whole point of most of these very urgent issues is to move to a system where the need for many, very repetitive page loads is completely eliminated in the first place.  That's the really fast way to speed up data entry.  #219 and #246 will have a huge impact on reducing the need for page loads, but all of the others at level 0 will also have significant impacts.  (Reapplying a previously applied filter often needs a long wait, too.  A 'Save' button that does not need a full reload would also help.)

#256 alone, for example, if implemented, will mean a reduction in the total number of page loads needed to encode a single language from about 220 to about 50.

Count that in seconds per page load, and you will see why those issues have the priority levels 0 and 1, and this one has level 4.  I doubt you'll be able to deliver a sudden 80% reduction in page load times, but #256 will give us an 80% reduction in page loads needed in the first place.

It's great that youve just managed a 35% improvement is load times (from 17 to 11 seconds on average is a big, big gain), and that it will work right across the board.  But I hope this comment puts that into perspective as against the other issues, and explains why this issue was down at priority rank 4.  Probably best to leave it for now and get on with the level 0 issues now, having already made a big improvement on this one for now, please!
